File: 1.两数之和.c
/*
 * @lc app=leetcode.cn id=1 lang=c
 *
 * [1] 两数之和
 */

// @lc code=start
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */

// 定义哈希表节点
typedef struct HashNode {
    int key; // 存放数组元素
    int value; // 存放数组下标
    struct HashNode* next;
} HashNode;

// 哈希函数
int hashFunction(int key, int size)
{
    return abs(key) % size; // abs()函数取绝对值
}

// 创建哈希表
HashNode** createHashTable(int size)
{
    HashNode** hashTable = (HashNode**)malloc(size * sizeof(HashNode*));

    // 初始化哈希表
    for (int i = 0; i < size; ++i) {
        hashTable[i] = NULL;
    }

    return hashTable;
}

// 插入哈希表
void insertHashTable(HashNode** hashTable, int key, int value, int size)
{
    int hashIndex = hashFunction(key, size);
    HashNode* newNode = (HashNode*)malloc(sizeof(HashNode));
    newNode->key = key;
    newNode->value = value;
    newNode->next = hashTable[hashIndex];
    hashTable[hashIndex] = newNode;
}

// 查找哈希表
HashNode* searchHashTable(HashNode** hashTable, int key, int size)
{
    int hashIndex = hashFunction(key, size);
    HashNode* node = hashTable[hashIndex];
    while (node != NULL) {
        if (node->key == key) {
            return node;
        }
        node = node->next;
    }
    return NULL;
}

// 释放哈希表
void freeHashTable(HashNode** hashTable, int size)
{
    for (int i = 0; i < size; ++i) {
        HashNode* current = hashTable[i];
        while (current != NULL) {
            HashNode* temp = current;
            current = current->next;
            free(temp);
        }
    }
    free(hashTable);
}

int* twoSum(int* nums, int numsSize, int target, int* returnSize)
{
    int hashTableSize = numsSize;
    HashNode** hashTable = createHashTable(hashTableSize);

    for (int i = 0; i < numsSize; ++i) {
        int complement = target - nums[i];
        HashNode* node = searchHashTable(hashTable, complement, hashTableSize);
        if (node != NULL) {
            int* results = (int*)malloc(2 * sizeof(int));
            results[0] = node->value;
            results[1] = i;
            *returnSize = 2;
            freeHashTable(hashTable, hashTableSize);
            return results;
        }
        insertHashTable(hashTable, nums[i], i, hashTableSize);
    }

    *returnSize = 0;
    freeHashTable(hashTable, hashTableSize);
    return NULL;
}

// @lc code=end

File: 1.两数之和.py
#
# @lc app=leetcode.cn id=1 lang=python3
#
# [1] 两数之和
#

# @lc code=start

# from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """ 解题思路：
        1. 使用哈希表存储数组元素及其索引
        2. 遍历数组，计算当前元素的补数（target - nums[i]）
        3. 检查补数是否在哈希表中
        4. 如果找到，返回当前索引和补数的索引
        5. 如果没有找到，将当前元素及其索引存入哈希表
        """
        num_to_index = {}
        for index, num in enumerate(nums):
            complement = target - num
            if complement in num_to_index:
                return [index, num_to_index[complement]]
            num_to_index[num] = index
        return []


# @lc code=end
